### ISSUE TITLE
Verify if internal data structure of spinetoolbox supports the expressiveness of rdf/rdfs/owl

### ISSUE THREAD

**Issue by joepvgenuchten** [Spine-project/Spine-Toolbox#2448]
There is a desire to be able to import RDF, see #2447 . We need to verify if the spinetoolbox data structure has sufficient expressiveness to support that.

the todo list would be:

- [x]  compare rdfs/owl to spinetoolbox data structure
- [ ]  create a mapping from rdfs/owl to spinetoolbox restricting us to existing structures and their intend
- [ ]  identify missing structures that need to be added to the spinetoolbox internal data structure
- [ ] add any missing structures that are required for processing rdf data

**Comment by jkiviluo** [comment]
When looking at the OWL ontology (https://www.w3.org/TR/2012/REC-owl2-primer-20121211/), there are couple of important sections. I've bolded them below:

- Section 4.1: This is basic for us (entity_class -- entity)
- _**Section 4.2**_: Class hierarchy is in our plan (Spine_data_structure_for_Mopo.drawio) and we have taken the first steps already with the abstract class. We need to activate the 'type' column in the entity_class__entity_class table. @manuelma would know more...
- Section 4.3 Class Disjointness: Could be a type in the previous (enforcing the disjointness in Toolbox is another question - I don't see a need for it in the short term)
- _**Section 4.4. Object Properties.**_ This is about establishing a predicate between a subject and an object. We can do this for relationship entities through a parameter, but there is no way to distinguish that the parameter is about a predicate (which could imply e.g. special visualization in the graph view or enforcement of logical rules in scripting). That could be accomplished with a hard-coded parameter name, but that would be ugly and prevent using that name for anything else. Parameter value list would enforce valid choices for predicates, but we really need a parameter type for something like 'enumerated_list_for_predicates'. That allows to build services on top of it.
- Section 4.5. Property hierarchies. This is about creating isMemberOf relation based on predicates (if you are a wife, then you're also a spouse). Not sure if we should implement this - would probably require a new table for parameter_definition__parameter_definition. At least not a high priority.
- Section 4.6. Domain and Range restrictions. Similar to above - could create logical enforment of class membership, but we don't need to worry about that (unless we really want to implement full OWL)
- Section 4.7. Equality and Inequality of individuals: We enforce unique names, so we don't need this functionality. However, we might need to resolve not-unique names that can come with OWL data. First implementation could just be to raise an error when uniqueness is violated.
- **Section 4.8: Datatypes.** Need to be checked that we can handle all OWL datatypes.
- Section 5 is about creating more complex logical rules - we don't need to do these unless we want to be able to resolve logics of OWL (let's just leave the responsibility to the user for now). Maybe one day Toolbox might want to be an OWL interface, but that's not in Mopo.
- Section 6. Advanced Use of Properties: These rules could be used to establish logic between methods. However, we can also do something more simple that suits our purposes and not try interpret this part of OWL (any data coming in from OWL or going to OWL is assumed to be ok)
- Section 7. Advanced Use of Datatypes: No need for these in Toolbox in Mopo.
- _**Section 8. Document Information and Annotations**_: Can have metadata and we should pick what is needed.
- Different OWL dialects - not sure if we need to dig into those. We will anyway implement only the part that makes sense.

**Comment by joepvgenuchten** [comment]
After some more thinking, I think the following constructs would make for a good MVP:

- rdfs:Class/owl:Class (we may treat them as equivalent)
- owl:ObjectProperty
- owl:DatatypeProperty
- rdf:List (as alternative way of expressing a high cardinality property)
- rdfs:subClassOf

**Comment by joepvgenuchten** [comment]
Yes, that is what I mean. I added the rdf:type construct, but I believe spine toolbox already has something for that.

When it comes to predicates, the question if these get their own table or if they become like join tables between (from and to) the entity table.

When it comes to the datatypes, we should consider whether we want to hold on to the  datatypes of the ingested data set or if we prefer to just transform them to the native (postgres? sqlite?) datatypes.
